File: console_client.py
import re
import socket
import threading
import time

class ConsoleClient:
    def __init__(self):
        self.nickname = ''
        self.server_ip = ''
        self.server_port = 0
        self.connected = False
        self.sock = None
        self.lock = threading.Lock()

    def input_nickname(self):
        pattern = re.compile(r'^[A-Za-z]+_[A-Za-z]+$')
        while True:
            name = input('Enter nickname (Name_Surname): ')
            if pattern.match(name):
                self.nickname = name
                break
            print('Invalid format. Try again.')

    def input_server(self):
        while True:
            addr = input('Enter server address (ip:port): ')
            if ':' in addr:
                ip, port = addr.split(':', 1)
                try:
                    port = int(port)
                    socket.inet_aton(ip)
                    self.server_ip = ip
                    self.server_port = port
                    break
                except (ValueError, OSError):
                    pass
            print('Invalid address. Try again.')

    def connect(self):
        with self.lock:
            if self.connected:
                print('Already connected.')
                return
            try:
                self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                self.sock.settimeout(5)
                self.sock.connect((self.server_ip, self.server_port))
                self.connected = True
                print('Connected to {}:{}.'.format(self.server_ip, self.server_port))
            except Exception as e:
                print('Failed to connect:', e)
                self.connected = False

    def disconnect(self):
        with self.lock:
            if self.sock:
                try:
                    self.sock.close()
                except Exception:
                    pass
            self.connected = False
            self.sock = None
            print('Disconnected.')

    def send_chat(self):
        if not self.connected:
            print('Not connected.')
            return
        while True:
            msg = input('Chat (0 to return): ')
            if msg == '0':
                break
            try:
                self.sock.sendall(msg.encode('utf-8'))
                print('Message sent.')
            except Exception as e:
                print('Failed to send:', e)
                self.disconnect()
                break

    def status_printer(self):
        while True:
            with self.lock:
                status = 'Connected' if self.connected else 'Disconnected'
            print('\n[{}]\n'.format(status))
            time.sleep(1)

    def run(self):
        self.input_nickname()
        self.input_server()
        threading.Thread(target=self.status_printer, daemon=True).start()
        while True:
            print('\n1. Connect')
            print('2. Disconnect')
            print('3. Chat')
            print('4. Change nickname')
            print('5. Change server')
            print('6. Quit')
            choice = input('Select option: ')
            if choice == '1':
                self.connect()
            elif choice == '2':
                self.disconnect()
            elif choice == '3':
                self.send_chat()
            elif choice == '4':
                self.input_nickname()
            elif choice == '5':
                self.input_server()
            elif choice == '6':
                self.disconnect()
                break
            else:
                print('Invalid option.')

if __name__ == '__main__':
    client = ConsoleClient()
    client.run()

File: client/console_client.cpp
#include <iostream>
#include <regex>
#include <thread>
#include <atomic>
#include <chrono>
#include <string>
#include "RakNetworkFactory.h"
#include "RakClientInterface.h"
#include "PacketEnumerations.h"
#include "BitStream.h"
#include "PacketPriority.h"
#include "SAMP/SAMPRPC.h"

using namespace RakNet;

static RakClientInterface *rakClient = nullptr;
static std::atomic<bool> connected{false};
static std::string nickname;
static std::string serverIp;
static unsigned short serverPort = 0;

bool validateNickname(const std::string &name)
{
    static const std::regex pattern("^[A-Za-z]+_[A-Za-z]+$");
    return std::regex_match(name, pattern);
}

bool parseAddress(const std::string &addr, std::string &ip, unsigned short &port)
{
    std::size_t pos = addr.find(':');
    if (pos == std::string::npos)
        return false;
    ip = addr.substr(0, pos);
    try
    {
        int p = std::stoi(addr.substr(pos + 1));
        if (p < 0 || p > 65535)
            return false;
        port = static_cast<unsigned short>(p);
    }
    catch (...)
    {
        return false;
    }
    return true;
}

void connectServer()
{
    if (connected)
    {
        std::cout << "Already connected.\n";
        return;
    }
    rakClient = RakNetworkFactory::GetRakClientInterface();
    if (!rakClient)
    {
        std::cout << "Failed to create client.\n";
        return;
    }
    rakClient->SetPassword("");
    if (rakClient->Connect(serverIp.c_str(), serverPort, 0, 0, 5))
    {
        connected = true;
        std::cout << "Connected to " << serverIp << ":" << serverPort << "\n";
    }
    else
    {
        std::cout << "Connection failed." << std::endl;
        RakNetworkFactory::DestroyRakClientInterface(rakClient);
        rakClient = nullptr;
    }
}

void disconnectServer()
{
    if (!connected)
        return;
    rakClient->Disconnect(0);
    RakNetworkFactory::DestroyRakClientInterface(rakClient);
    rakClient = nullptr;
    connected = false;
    std::cout << "Disconnected." << std::endl;
}

void sendChat(const std::string &text)
{
    if (!connected)
    {
        std::cout << "Not connected." << std::endl;
        return;
    }
    BitStream bs;
    unsigned char len = static_cast<unsigned char>(text.size());
    bs.Write(len);
    bs.Write(text.c_str(), len);
    rakClient->RPC(&RPC_Chat, &bs, HIGH_PRIORITY, RELIABLE, 0, FALSE, UNASSIGNED_NETWORK_ID, nullptr);
    std::cout << "Message sent." << std::endl;
}

void statusPrinter()
{
    while (true)
    {
        std::cout << "\n[" << (connected ? "Connected" : "Disconnected") << "]\n";
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

void inputNickname()
{
    std::string name;
    do
    {
        std::cout << "Enter nickname (Name_Surname): ";
        std::getline(std::cin, name);
    } while (!validateNickname(name));
    nickname = name;
}

void inputServer()
{
    std::string addr;
    do
    {
        std::cout << "Enter server address (ip:port): ";
        std::getline(std::cin, addr);
    } while (!parseAddress(addr, serverIp, serverPort));
}

int main()
{
    inputNickname();
    inputServer();

    std::thread(statusPrinter).detach();
    while (true)
    {
        std::cout << "\n1. Connect\n2. Disconnect\n3. Chat\n4. Change nickname\n5. Change server\n6. Quit\nSelect option: ";
        std::string choice;
        std::getline(std::cin, choice);
        if (choice == "1")
        {
            connectServer();
        }
        else if (choice == "2")
        {
            disconnectServer();
        }
        else if (choice == "3")
        {
            while (true)
            {
                std::cout << "Chat (0 to return): ";
                std::string msg;
                std::getline(std::cin, msg);
                if (msg == "0")
                    break;
                sendChat(msg);
            }
        }
        else if (choice == "4")
        {
            inputNickname();
        }
        else if (choice == "5")
        {
            inputServer();
        }
        else if (choice == "6")
        {
            disconnectServer();
            break;
        }
        else
        {
            std::cout << "Invalid option." << std::endl;
        }
    }
    return 0;
}

